File: go/internal/application/place_order.go
package application

import (
	"context"
	"fmt"
	"strings"

	"warehouse/internal/domain"
)

// PlaceOrderResult is the typed outcome of the place-order use case: no
// exception ever crosses the boundary. Exactly one field is meaningful —
// Order is non-nil on success, Failure carries the wrapped domain error on
// failure.
type PlaceOrderResult struct {
	Order   *domain.Order
	Failure error
}

// Failed reports whether the use case ended in a typed failure.
func (r PlaceOrderResult) Failed() bool {
	return r.Failure != nil
}

// PlaceOrderUseCase orchestrates validate -> reserveAll -> charge -> pay ->
// persist, compensating on decline or save failure.
type PlaceOrderUseCase struct {
	inventory  InventoryGateway
	payments   PaymentProcessor
	repository OrderRepository
	ids        OrderIDGenerator
}

// NewPlaceOrderUseCase wires the use case to its outbound ports.
func NewPlaceOrderUseCase(
	inventory InventoryGateway,
	payments PaymentProcessor,
	repository OrderRepository,
	ids OrderIDGenerator,
) *PlaceOrderUseCase {
	return &PlaceOrderUseCase{
		inventory:  inventory,
		payments:   payments,
		repository: repository,
		ids:        ids,
	}
}

// Execute validates the order, atomically reserves stock, collects payment,
// marks the order PAID, then persists. Declined charges release the
// reservation. Save failures refund and release. The same idempotency key
// and payload replays the original snapshot without a second charge.
func (uc *PlaceOrderUseCase) Execute(ctx context.Context, lines []domain.OrderLine, idempotencyKey string) PlaceOrderResult {
	fingerprint := fingerprintLines(lines)
	if result, done := uc.replay(ctx, idempotencyKey, fingerprint); done {
		return result
	}
	order, err := domain.NewOrder(lines, uc.ids.Next())
	if err != nil {
		return PlaceOrderResult{Order: nil, Failure: fmt.Errorf("place order: %w", err)}
	}
	token, err := uc.inventory.ReserveAll(ctx, order.ID(), order.Lines(), idempotencyKey)
	if err != nil {
		return PlaceOrderResult{Order: nil, Failure: fmt.Errorf("reserve stock: %w", err)}
	}
	receipt, err := uc.payments.Charge(ctx, order, idempotencyKey)
	if err != nil {
		return uc.releaseAfterDecline(ctx, token, fmt.Errorf("charge payment: %w", err))
	}
	if err := order.Pay(); err != nil {
		return uc.compensate(ctx, token, receipt, fmt.Errorf("pay order: %w", err))
	}
	saved, err := uc.repository.Save(ctx, order, order.Version())
	if err != nil {
		return uc.compensate(ctx, token, receipt, fmt.Errorf("persist order: %w", err))
	}
	uc.repository.RememberIdempotency(ctx, idempotencyKey, fingerprint, saved)
	return PlaceOrderResult{Order: saved, Failure: nil}
}

func (uc *PlaceOrderUseCase) replay(ctx context.Context, key, fingerprint string) (PlaceOrderResult, bool) {
	prior, snapshot, ok := uc.repository.GetByIdempotencyKey(ctx, key)
	if !ok {
		return PlaceOrderResult{Order: nil, Failure: nil}, false
	}
	if prior != fingerprint {
		return PlaceOrderResult{
			Order:   nil,
			Failure: fmt.Errorf("place order: %w", domain.InvalidOrder{Reason: "idempotency key reused with different payload"}),
		}, true
	}
	return PlaceOrderResult{Order: snapshot, Failure: nil}, true
}

func (uc *PlaceOrderUseCase) releaseAfterDecline(ctx context.Context, token ReservationToken, cause error) PlaceOrderResult {
	if err := uc.inventory.Release(ctx, token); err != nil {
		return PlaceOrderResult{Order: nil, Failure: fmt.Errorf("release reservation: %w", err)}
	}
	return PlaceOrderResult{Order: nil, Failure: cause}
}

func (uc *PlaceOrderUseCase) compensate(ctx context.Context, token ReservationToken, receipt ChargeReceipt, cause error) PlaceOrderResult {
	if err := uc.payments.Refund(ctx, receipt); err != nil {
		return PlaceOrderResult{Order: nil, Failure: fmt.Errorf("compensate refund: %w", err)}
	}
	if err := uc.inventory.Release(ctx, token); err != nil {
		return PlaceOrderResult{Order: nil, Failure: fmt.Errorf("compensate release: %w", err)}
	}
	return PlaceOrderResult{Order: nil, Failure: cause}
}

func fingerprintLines(lines []domain.OrderLine) string {
	parts := make([]string, 0, len(lines))
	for _, line := range lines {
		parts = append(parts, fmt.Sprintf(
			"%s:%d:%s:%d",
			string(line.SKU),
			int64(line.Quantity),
			line.UnitPrice.Currency,
			line.UnitPrice.MinorUnits,
		))
	}
	return strings.Join(parts, "|")
}

File: csharp/src/Warehouse.Application/PlaceOrder.cs
using System.Globalization;
using System.Text;
using Warehouse.Domain;

namespace Warehouse.Application;

public abstract record PlaceOrderResult
{
    public sealed record Success(Order Order) : PlaceOrderResult;

    public sealed record Failure(DomainException Error) : PlaceOrderResult;
}

public sealed class PlaceOrderUseCase
{
    private readonly IInventoryGateway _inventory;
    private readonly IPaymentProcessor _payments;
    private readonly IOrderRepository _repository;
    private readonly IOrderIdGenerator _ids;

    public PlaceOrderUseCase(
        IInventoryGateway inventory,
        IPaymentProcessor payments,
        IOrderRepository repository,
        IOrderIdGenerator ids)
    {
        ArgumentNullException.ThrowIfNull(inventory);
        ArgumentNullException.ThrowIfNull(payments);
        ArgumentNullException.ThrowIfNull(repository);
        ArgumentNullException.ThrowIfNull(ids);
        _inventory = inventory;
        _payments = payments;
        _repository = repository;
        _ids = ids;
    }

    public PlaceOrderResult Execute(IReadOnlyList<OrderLine> lines, string idempotencyKey)
    {
        ArgumentNullException.ThrowIfNull(lines);
        ArgumentNullException.ThrowIfNull(idempotencyKey);
        string fingerprint = Fingerprint(lines);
        IdempotencyRecord? remembered = _repository.GetByIdempotencyKey(idempotencyKey);
        if (remembered is { } prior)
        {
            if (prior.Fingerprint != fingerprint)
            {
                return new PlaceOrderResult.Failure(
                    new InvalidOrderException("idempotency key reused with different payload"));
            }

            return new PlaceOrderResult.Success(prior.Order);
        }

        Order order;
        try
        {
            order = new Order(_ids.Next(), lines);
        }
        catch (InvalidOrderException error)
        {
            return new PlaceOrderResult.Failure(error);
        }

        ReservationResult reserved = _inventory.ReserveAll(order.Id, order.Lines, idempotencyKey);
        if (reserved is ReservationResult.Shortage shortage)
        {
            return new PlaceOrderResult.Failure(shortage.Error);
        }

        var reservedOk = (ReservationResult.Reserved)reserved;
        ChargeResult charged = _payments.Charge(order, idempotencyKey);
        if (charged is ChargeResult.Declined declined)
        {
            return ReleaseOrFail(reservedOk.Token, declined.Error);
        }

        var chargedOk = (ChargeResult.Charged)charged;
        try
        {
            order.Pay();
        }
        catch (InvalidOrderException error)
        {
            return Compensate(reservedOk.Token, chargedOk.Receipt, error);
        }

        SaveResult saved = _repository.Save(order, expectedVersion: 0);
        if (saved is SaveResult.Conflict conflict)
        {
            return Compensate(reservedOk.Token, chargedOk.Receipt, conflict.Error);
        }

        var savedOk = (SaveResult.Saved)saved;
        _repository.RememberIdempotency(idempotencyKey, fingerprint, savedOk.Snapshot);
        return new PlaceOrderResult.Success(savedOk.Snapshot);
    }

    private PlaceOrderResult ReleaseOrFail(ReservationToken token, DomainException error)
    {
        ReleaseResult released = _inventory.Release(token);
        if (released is ReleaseResult.Failed failed)
        {
            return new PlaceOrderResult.Failure(failed.Error);
        }

        return new PlaceOrderResult.Failure(error);
    }

    private PlaceOrderResult Compensate(ReservationToken token, ChargeReceipt receipt, DomainException error)
    {
        RefundResult refunded = _payments.Refund(receipt);
        if (refunded is RefundResult.Failed failedRefund)
        {
            return new PlaceOrderResult.Failure(failedRefund.Error);
        }

        return ReleaseOrFail(token, error);
    }

    private static string Fingerprint(IReadOnlyList<OrderLine> lines)
    {
        var builder = new StringBuilder();
        for (int i = 0; i < lines.Count; i++)
        {
            if (i > 0)
            {
                builder.Append('|');
            }

            OrderLine line = lines[i];
            builder.Append(line.Sku.Code);
            builder.Append(':');
            builder.Append(line.Quantity.Value.ToString(CultureInfo.InvariantCulture));
            builder.Append(':');
            builder.Append(line.UnitPrice.Currency);
            builder.Append(':');
            builder.Append(line.UnitPrice.MinorUnits.ToString(CultureInfo.InvariantCulture));
        }

        return builder.ToString();
    }
}
